streams/src/main/java/org/apache/kafka/streams/errors/SerializationExceptionHandler.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.errors;
+
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.common.Configurable;
+import org.apache.kafka.common.serialization.Serializer;
+
+/**
+ * Interface that specifies how an exception while attempting to serialize the
+ * the record should be handled
+ */
+public interface SerializationExceptionHandler extends Configurable {
```

Comment:
This is a public API change and would require a KIP.

Why do we need this class? Can't we use the existing `ProductionExceptionHandler` as-is?

Reply:
I need a new handler to expose the `serializer` which throws error and to specially tell the user that it happened when serializing the record key / value. I ll write the KIP.

Reply:
I cannot follow?

> to expose the serializer which throws error

Why do we need to provide the serializer into the handler?

> and to specially tell the user that it happened when serializing the record key / value

We can use two try-catch-blocks, one for calling the key-serializer and one for the value-serializer to do this if we need to.

Reply:
ok, I thought it might be useful for the user to log the serializer class type - if it's not useful I'm ok to remove it. 

While implementing, I noticed that the existing method in `ProductionExceptionHandler#handle` accepts the `ProducerRecord` with key and value type as `byte[]`. On serialization exception, we can send a generic record. But, it misleads the user as the key and value type is **not** `byte[]`.

Reply:
That's a good point... Maybe it's worth a KIP to fix the type issue. I was hoping to do a simple fix, but it seems this would not result in a clean solution and doing a KIP might be better. Thanks for clarifying!

Reply:
If all you want to do is report the serializer type, it seems like we can add it to the error message without introducing a new exception type. Is that right, or have I missed something?

Reply:
@vvcephei 
I haven't updated this PR. I'm agreed to remove the serializer. I waiting for the [KIP-399](https://cwiki.apache.org/confluence/display/KAFKA/KIP-399%3A+Extend+ProductionExceptionHandler+to+cover+serialization+exceptions) to pass.
Note the issue here is the existing `handle` method has `ProducerRecord` in which `key` and `value` type is set as `byte[]`. So, I require a new `handle` method for backward compatibility.

Reply:
@mjsax @vvcephei 
Updated the PR and the [KIP-399](https://cwiki.apache.org/confluence/display/KAFKA/KIP-399%3A+Extend+ProductionExceptionHandler+to+cover+serialization+exceptions). Please take another look.